File: crates/iscc-lib/src/conformance.rs
//! Conformance selftest for ISO 24138:2024 (ISCC).
//!
//! Runs all 9 gen functions against vendored conformance vectors from `data.json`
//! and reports pass/fail. An application that claims ISCC conformance MUST pass
//! all tests in this suite.

use crate::{
    gen_audio_code_v0, gen_data_code_v0, gen_image_code_v0, gen_instance_code_v0, gen_iscc_code_v0,
    gen_meta_code_v0, gen_mixed_code_v0, gen_text_code_v0, gen_video_code_v0,
};

/// Embedded conformance test vectors (compile-time).
const TEST_DATA: &str = include_str!("../tests/data.json");

/// Run all conformance tests against vendored test vectors.
///
/// Iterates through all 9 `gen_*_v0` function sections in the conformance data,
/// calls each function with the specified inputs, and compares the `.iscc` field
/// of the result against expected output. Returns `true` if all tests pass,
/// `false` if any mismatch or error occurs. Does not panic — logs failures via
/// `eprintln!` and continues through all test cases.
pub fn conformance_selftest() -> bool {
    let data: serde_json::Value = match serde_json::from_str(TEST_DATA) {
        Ok(v) => v,
        Err(e) => {
            eprintln!("FAILED: could not parse conformance data: {e}");
            return false;
        }
    };

    let mut passed = true;

    passed &= run_meta_tests(&data);
    passed &= run_text_tests(&data);
    passed &= run_image_tests(&data);
    passed &= run_audio_tests(&data);
    passed &= run_video_tests(&data);
    passed &= run_mixed_tests(&data);
    passed &= run_data_tests(&data);
    passed &= run_instance_tests(&data);
    passed &= run_iscc_tests(&data);

    passed
}

/// Run conformance tests for `gen_meta_code_v0`.
fn run_meta_tests(data: &serde_json::Value) -> bool {
    let mut passed = true;
    let section = &data["gen_meta_code_v0"];
    let cases = match section.as_object() {
        Some(c) => c,
        None => {
            eprintln!("FAILED: gen_meta_code_v0 section missing from conformance data");
            return false;
        }
    };

    for (tc_name, tc) in cases {
        let func_name = "gen_meta_code_v0";
        let result = (|| {
            let inputs = tc["inputs"].as_array()?;
            let name = inputs[0].as_str()?;
            let desc_str = inputs[1].as_str()?;
            let meta_val = &inputs[2];
            let bits = inputs[3].as_u64()? as u32;

            let meta_arg: Option<String> = match meta_val {
                serde_json::Value::Null => None,
                serde_json::Value::String(s) => Some(s.clone()),
                serde_json::Value::Object(_) => serde_json::to_string(meta_val).ok(),
                _ => None,
            };

            let desc = if desc_str.is_empty() {
                None
            } else {
                Some(desc_str)
            };

            let expected_iscc = tc["outputs"]["iscc"].as_str()?;
            match gen_meta_code_v0(name, desc, meta_arg.as_deref(), bits) {
                Ok(result) if result.iscc == expected_iscc => Some(true),
                Ok(result) => {
                    eprintln!(
                        "FAILED: {func_name}.{tc_name} — expected {expected_iscc}, got {}",
                        result.iscc
                    );
                    Some(false)
                }
                Err(e) => {
                    eprintln!("FAILED: {func_name}.{tc_name} — error: {e}");
                    Some(false)
                }
            }
        })();

        match result {
            Some(true) => {}
            Some(false) => passed = false,
            None => {
                eprintln!("FAILED: {func_name}.{tc_name} — could not parse test inputs");
                passed = false;
            }
        }
    }
    passed
}

/// Run conformance tests for `gen_text_code_v0`.
fn run_text_tests(data: &serde_json::Value) -> bool {
    let mut passed = true;
    let section = &data["gen_text_code_v0"];
    let cases = match section.as_object() {
        Some(c) => c,
        None => {
            eprintln!("FAILED: gen_text_code_v0 section missing from conformance data");
            return false;
        }
    };

    for (tc_name, tc) in cases {
        let func_name = "gen_text_code_v0";
        let result = (|| {
            let inputs = tc["inputs"].as_array()?;
            let text = inputs[0].as_str()?;
            let bits = inputs[1].as_u64()? as u32;
            let expected_iscc = tc["outputs"]["iscc"].as_str()?;

            match gen_text_code_v0(text, bits) {
                Ok(result) if result.iscc == expected_iscc => Some(true),
                Ok(result) => {
                    eprintln!(
                        "FAILED: {func_name}.{tc_name} — expected {expected_iscc}, got {}",
                        result.iscc
                    );
                    Some(false)
                }
                Err(e) => {
                    eprintln!("FAILED: {func_name}.{tc_name} — error: {e}");
                    Some(false)
                }
            }
        })();

        match result {
            Some(true) => {}
            Some(false) => passed = false,
            None => {
                eprintln!("FAILED: {func_name}.{tc_name} — could not parse test inputs");
                passed = false;
            }
        }
    }
    passed
}

/// Run conformance tests for `gen_image_code_v0`.
fn run_image_tests(data: &serde_json::Value) -> bool {
    let mut passed = true;
    let section = &data["gen_image_code_v0"];
    let cases = match section.as_object() {
        Some(c) => c,
        None => {
            eprintln!("FAILED: gen_image_code_v0 section missing from conformance data");
            return false;
        }
    };

    for (tc_name, tc) in cases {
        let func_name = "gen_image_code_v0";
        let result = (|| {
            let inputs = tc["inputs"].as_array()?;
            let pixels_json = inputs[0].as_array()?;
            let bits = inputs[1].as_u64()? as u32;
            let expected_iscc = tc["outputs"]["iscc"].as_str()?;

            let pixels: Vec<u8> = pixels_json
                .iter()
                .map(|v| v.as_u64().map(|n| n as u8))
                .collect::<Option<Vec<u8>>>()?;

            match gen_image_code_v0(&pixels, bits) {
                Ok(result) if result.iscc == expected_iscc => Some(true),
                Ok(result) => {
                    eprintln!(
                        "FAILED: {func_name}.{tc_name} — expected {expected_iscc}, got {}",
                        result.iscc
                    );
                    Some(false)
                }
                Err(e) => {
                    eprintln!("FAILED: {func_name}.{tc_name} — error: {e}");
                    Some(false)
                }
            }
        })();

        match result {
            Some(true) => {}
            Some(false) => passed = false,
            None => {
                eprintln!("FAILED: {func_name}.{tc_name} — could not parse test inputs");
                passed = false;
            }
        }
    }
    passed
}

/// Run conformance tests for `gen_audio_code_v0`.
fn run_audio_tests(data: &serde_json::Value) -> bool {
    let mut passed = true;
    let section = &data["gen_audio_code_v0"];
    let cases = match section.as_object() {
        Some(c) => c,
        None => {
            eprintln!("FAILED: gen_audio_code_v0 section missing from conformance data");
            return false;
        }
    };

    for (tc_name, tc) in cases {
        let func_name = "gen_audio_code_v0";
        let result = (|| {
            let inputs = tc["inputs"].as_array()?;
            let cv_json = inputs[0].as_array()?;
            let bits = inputs[1].as_u64()? as u32;
            let expected_iscc = tc["outputs"]["iscc"].as_str()?;

            let cv: Vec<i32> = cv_json
                .iter()
                .map(|v| v.as_i64().map(|n| n as i32))
                .collect::<Option<Vec<i32>>>()?;

            match gen_audio_code_v0(&cv, bits) {
                Ok(result) if result.iscc == expected_iscc => Some(true),
                Ok(result) => {
                    eprintln!(
                        "FAILED: {func_name}.{tc_name} — expected {expected_iscc}, got {}",
                        result.iscc
                    );
                    Some(false)
                }
                Err(e) => {
                    eprintln!("FAILED: {func_name}.{tc_name} — error: {e}");
                    Some(false)
                }
            }
        })();

        match result {
            Some(true) => {}
            Some(false) => passed = false,
            None => {
                eprintln!("FAILED: {func_name}.{tc_name} — could not parse test inputs");
                passed = false;
            }
        }
    }
    passed
}

/// Run conformance tests for `gen_video_code_v0`.
fn run_video_tests(data: &serde_json::Value) -> bool {
    let mut passed = true;
    let section = &data["gen_video_code_v0"];
    let cases = match section.as_object() {
        Some(c) => c,
        None => {
            eprintln!("FAILED: gen_video_code_v0 section missing from conformance data");
            return false;
        }
    };

    for (tc_name, tc) in cases {
        let func_name = "gen_video_code_v0";
        let result = (|| {
            let inputs = tc["inputs"].as_array()?;
            let frames_json = inputs[0].as_array()?;
            let bits = inputs[1].as_u64()? as u32;
            let expected_iscc = tc["outputs"]["iscc"].as_str()?;

            let frame_sigs: Option<Vec<Vec<i32>>> = frames_json
                .iter()
                .map(|frame| {
                    frame
                        .as_array()?
                        .iter()
                        .map(|v| v.as_i64().map(|n| n as i32))
                        .collect::<Option<Vec<i32>>>()
                })
                .collect();
            let frame_sigs = frame_sigs?;

            match gen_video_code_v0(&frame_sigs, bits) {
                Ok(result) if result.iscc == expected_iscc => Some(true),
                Ok(result) => {
                    eprintln!(
                        "FAILED: {func_name}.{tc_name} — expected {expected_iscc}, got {}",
                        result.iscc
                    );
                    Some(false)
                }
                Err(e) => {
                    eprintln!("FAILED: {func_name}.{tc_name} — error: {e}");
                    Some(false)
                }
            }
        })();

        match result {
            Some(true) => {}
            Some(false) => passed = false,
            None => {
                eprintln!("FAILED: {func_name}.{tc_name} — could not parse test inputs");
                passed = false;
            }
        }
    }
    passed
}

/// Run conformance tests for `gen_mixed_code_v0`.
fn run_mixed_tests(data: &serde_json::Value) -> bool {
    let mut passed = true;
    let section = &data["gen_mixed_code_v0"];
    let cases = match section.as_object() {
        Some(c) => c,
        None => {
            eprintln!("FAILED: gen_mixed_code_v0 section missing from conformance data");
            return false;
        }
    };

    for (tc_name, tc) in cases {
        let func_name = "gen_mixed_code_v0";
        let result = (|| {
            let inputs = tc["inputs"].as_array()?;
            let codes_json = inputs[0].as_array()?;
            let bits = inputs[1].as_u64()? as u32;
            let expected_iscc = tc["outputs"]["iscc"].as_str()?;

            let codes_owned: Vec<String> = codes_json
                .iter()
                .filter_map(|v| v.as_str().map(String::from))
                .collect();
            let codes: Vec<&str> = codes_owned.iter().map(|s| s.as_str()).collect();

            match gen_mixed_code_v0(&codes, bits) {
                Ok(result) if result.iscc == expected_iscc => Some(true),
                Ok(result) => {
                    eprintln!(
                        "FAILED: {func_name}.{tc_name} — expected {expected_iscc}, got {}",
                        result.iscc
                    );
                    Some(false)
                }
                Err(e) => {
                    eprintln!("FAILED: {func_name}.{tc_name} — error: {e}");
                    Some(false)
                }
            }
        })();

        match result {
            Some(true) => {}
            Some(false) => passed = false,
            None => {
                eprintln!("FAILED: {func_name}.{tc_name} — could not parse test inputs");
                passed = false;
            }
        }
    }
    passed
}

/// Decode a `"stream:<hex>"` value into bytes.
fn decode_stream(s: &str) -> Option<Vec<u8>> {
    let hex_data = s.strip_prefix("stream:")?;
    hex::decode(hex_data).ok()
}

/// Run conformance tests for `gen_data_code_v0`.
fn run_data_tests(data: &serde_json::Value) -> bool {
    let mut passed = true;
    let section = &data["gen_data_code_v0"];
    let cases = match section.as_object() {
        Some(c) => c,
        None => {
            eprintln!("FAILED: gen_data_code_v0 section missing from conformance data");
            return false;
        }
    };

    for (tc_name, tc) in cases {
        let func_name = "gen_data_code_v0";
        let result = (|| {
            let inputs = tc["inputs"].as_array()?;
            let stream_str = inputs[0].as_str()?;
            let bits = inputs[1].as_u64()? as u32;
            let expected_iscc = tc["outputs"]["iscc"].as_str()?;
            let input_bytes = decode_stream(stream_str)?;

            match gen_data_code_v0(&input_bytes, bits) {
                Ok(result) if result.iscc == expected_iscc => Some(true),
                Ok(result) => {
                    eprintln!(
                        "FAILED: {func_name}.{tc_name} — expected {expected_iscc}, got {}",
                        result.iscc
                    );
                    Some(false)
                }
                Err(e) => {
                    eprintln!("FAILED: {func_name}.{tc_name} — error: {e}");
                    Some(false)
                }
            }
        })();

        match result {
            Some(true) => {}
            Some(false) => passed = false,
            None => {
                eprintln!("FAILED: {func_name}.{tc_name} — could not parse test inputs");
                passed = false;
            }
        }
    }
    passed
}

/// Run conformance tests for `gen_instance_code_v0`.
fn run_instance_tests(data: &serde_json::Value) -> bool {
    let mut passed = true;
    let section = &data["gen_instance_code_v0"];
    let cases = match section.as_object() {
        Some(c) => c,
        None => {
            eprintln!("FAILED: gen_instance_code_v0 section missing from conformance data");
            return false;
        }
    };

    for (tc_name, tc) in cases {
        let func_name = "gen_instance_code_v0";
        let result = (|| {
            let inputs = tc["inputs"].as_array()?;
            let stream_str = inputs[0].as_str()?;
            let bits = inputs[1].as_u64()? as u32;
            let expected_iscc = tc["outputs"]["iscc"].as_str()?;
            let input_bytes = decode_stream(stream_str)?;

            match gen_instance_code_v0(&input_bytes, bits) {
                Ok(result) if result.iscc == expected_iscc => Some(true),
                Ok(result) => {
                    eprintln!(
                        "FAILED: {func_name}.{tc_name} — expected {expected_iscc}, got {}",
                        result.iscc
                    );
                    Some(false)
                }
                Err(e) => {
                    eprintln!("FAILED: {func_name}.{tc_name} — error: {e}");
                    Some(false)
                }
            }
        })();

        match result {
            Some(true) => {}
            Some(false) => passed = false,
            None => {
                eprintln!("FAILED: {func_name}.{tc_name} — could not parse test inputs");
                passed = false;
            }
        }
    }
    passed
}

/// Run conformance tests for `gen_iscc_code_v0`.
fn run_iscc_tests(data: &serde_json::Value) -> bool {
    let mut passed = true;
    let section = &data["gen_iscc_code_v0"];
    let cases = match section.as_object() {
        Some(c) => c,
        None => {
            eprintln!("FAILED: gen_iscc_code_v0 section missing from conformance data");
            return false;
        }
    };

    for (tc_name, tc) in cases {
        let func_name = "gen_iscc_code_v0";
        let result = (|| {
            let inputs = tc["inputs"].as_array()?;
            let codes_json = inputs[0].as_array()?;
            let expected_iscc = tc["outputs"]["iscc"].as_str()?;

            let codes_owned: Vec<String> = codes_json
                .iter()
                .filter_map(|v| v.as_str().map(String::from))
                .collect();
            let codes: Vec<&str> = codes_owned.iter().map(|s| s.as_str()).collect();

            // Conformance vectors use default (non-wide) mode
            match gen_iscc_code_v0(&codes, false) {
                Ok(result) if result.iscc == expected_iscc => Some(true),
                Ok(result) => {
                    eprintln!(
                        "FAILED: {func_name}.{tc_name} — expected {expected_iscc}, got {}",
                        result.iscc
                    );
                    Some(false)
                }
                Err(e) => {
                    eprintln!("FAILED: {func_name}.{tc_name} — error: {e}");
                    Some(false)
                }
            }
        })();

        match result {
            Some(true) => {}
            Some(false) => passed = false,
            None => {
                eprintln!("FAILED: {func_name}.{tc_name} — could not parse test inputs");
                passed = false;
            }
        }
    }
    passed
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_conformance_selftest_passes() {
        assert!(conformance_selftest());
    }
}

File: packages/go/conformance.go
// Conformance selftest for ISO 24138:2024 (ISCC).
// Runs all 9 gen functions against vendored conformance vectors from data.json
// and reports pass/fail. An application that claims ISCC conformance MUST pass
// all tests in this suite.
package iscc

import (
	_ "embed"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"os"
	"strings"
)

//go:embed testdata/data.json
var conformanceData string

// vectorEntry holds a single conformance test case from data.json.
type vectorEntry struct {
	Inputs  []json.RawMessage     `json:"inputs"`
	Outputs map[string]interface{} `json:"outputs"`
}

// ConformanceSelftest validates all 46 conformance vectors from data.json.
// Returns (true, nil) if all vectors pass, (false, nil) if any mismatch,
// and (false, error) if data.json cannot be parsed.
func ConformanceSelftest() (bool, error) {
	var data map[string]map[string]vectorEntry
	if err := json.Unmarshal([]byte(conformanceData), &data); err != nil {
		return false, fmt.Errorf("could not parse conformance data: %w", err)
	}

	passed := true
	passed = runMetaTests(data) && passed
	passed = runTextTests(data) && passed
	passed = runImageTests(data) && passed
	passed = runAudioTests(data) && passed
	passed = runVideoTests(data) && passed
	passed = runMixedTests(data) && passed
	passed = runDataTests(data) && passed
	passed = runInstanceTests(data) && passed
	passed = runIsccTests(data) && passed

	return passed, nil
}

// runMetaTests validates gen_meta_code_v0 conformance vectors.
func runMetaTests(data map[string]map[string]vectorEntry) bool {
	funcName := "gen_meta_code_v0"
	cases, ok := data[funcName]
	if !ok {
		fmt.Fprintf(os.Stderr, "FAILED: %s section missing from conformance data\n", funcName)
		return false
	}

	passed := true
	for tcName, tc := range cases {
		if !runMetaCase(funcName, tcName, tc) {
			passed = false
		}
	}
	return passed
}

// runMetaCase runs a single gen_meta_code_v0 test case.
func runMetaCase(funcName, tcName string, tc vectorEntry) bool {
	var inputName string
	if err := json.Unmarshal(tc.Inputs[0], &inputName); err != nil {
		fmt.Fprintf(os.Stderr, "FAILED: %s.%s — could not parse test inputs\n", funcName, tcName)
		return false
	}

	// Description: string or null
	var desc *string
	var descStr string
	if string(tc.Inputs[1]) != "null" {
		if err := json.Unmarshal(tc.Inputs[1], &descStr); err != nil {
			fmt.Fprintf(os.Stderr, "FAILED: %s.%s — could not parse test inputs\n", funcName, tcName)
			return false
		}
		if descStr != "" {
			desc = &descStr
		}
	}

	// Meta: null, string, or JSON object
	var meta *string
	rawMeta := string(tc.Inputs[2])
	if rawMeta != "null" {
		if rawMeta[0] == '"' {
			var metaStr string
			if err := json.Unmarshal(tc.Inputs[2], &metaStr); err != nil {
				fmt.Fprintf(os.Stderr, "FAILED: %s.%s — could not parse test inputs\n", funcName, tcName)
				return false
			}
			meta = &metaStr
		} else {
			metaStr := rawMeta
			meta = &metaStr
		}
	}

	var bits float64
	if err := json.Unmarshal(tc.Inputs[3], &bits); err != nil {
		fmt.Fprintf(os.Stderr, "FAILED: %s.%s — could not parse test inputs\n", funcName, tcName)
		return false
	}

	expectedISCC, _ := tc.Outputs["iscc"].(string)
	result, err := GenMetaCodeV0(inputName, desc, meta, uint32(bits))
	if err != nil {
		fmt.Fprintf(os.Stderr, "FAILED: %s.%s — error: %v\n", funcName, tcName, err)
		return false
	}
	if result.Iscc != expectedISCC {
		fmt.Fprintf(os.Stderr, "FAILED: %s.%s — expected %s, got %s\n", funcName, tcName, expectedISCC, result.Iscc)
		return false
	}
	return true
}

// runTextTests validates gen_text_code_v0 conformance vectors.
func runTextTests(data map[string]map[string]vectorEntry) bool {
	funcName := "gen_text_code_v0"
	cases, ok := data[funcName]
	if !ok {
		fmt.Fprintf(os.Stderr, "FAILED: %s section missing from conformance data\n", funcName)
		return false
	}

	passed := true
	for tcName, tc := range cases {
		var text string
		if err := json.Unmarshal(tc.Inputs[0], &text); err != nil {
			fmt.Fprintf(os.Stderr, "FAILED: %s.%s — could not parse test inputs\n", funcName, tcName)
			passed = false
			continue
		}

		var bits float64
		if err := json.Unmarshal(tc.Inputs[1], &bits); err != nil {
			fmt.Fprintf(os.Stderr, "FAILED: %s.%s — could not parse test inputs\n", funcName, tcName)
			passed = false
			continue
		}

		expectedISCC, _ := tc.Outputs["iscc"].(string)
		result, err := GenTextCodeV0(text, uint32(bits))
		if err != nil {
			fmt.Fprintf(os.Stderr, "FAILED: %s.%s — error: %v\n", funcName, tcName, err)
			passed = false
			continue
		}
		if result.Iscc != expectedISCC {
			fmt.Fprintf(os.Stderr, "FAILED: %s.%s — expected %s, got %s\n", funcName, tcName, expectedISCC, result.Iscc)
			passed = false
		}
	}
	return passed
}

// runImageTests validates gen_image_code_v0 conformance vectors.
func runImageTests(data map[string]map[string]vectorEntry) bool {
	funcName := "gen_image_code_v0"
	cases, ok := data[funcName]
	if !ok {
		fmt.Fprintf(os.Stderr, "FAILED: %s section missing from conformance data\n", funcName)
		return false
	}

	passed := true
	for tcName, tc := range cases {
		var pixelsJSON []float64
		if err := json.Unmarshal(tc.Inputs[0], &pixelsJSON); err != nil {
			fmt.Fprintf(os.Stderr, "FAILED: %s.%s — could not parse test inputs\n", funcName, tcName)
			passed = false
			continue
		}
		pixels := make([]byte, len(pixelsJSON))
		for i, v := range pixelsJSON {
			pixels[i] = byte(v)
		}

		var bits float64
		if err := json.Unmarshal(tc.Inputs[1], &bits); err != nil {
			fmt.Fprintf(os.Stderr, "FAILED: %s.%s — could not parse test inputs\n", funcName, tcName)
			passed = false
			continue
		}

		expectedISCC, _ := tc.Outputs["iscc"].(string)
		result, err := GenImageCodeV0(pixels, uint32(bits))
		if err != nil {
			fmt.Fprintf(os.Stderr, "FAILED: %s.%s — error: %v\n", funcName, tcName, err)
			passed = false
			continue
		}
		if result.Iscc != expectedISCC {
			fmt.Fprintf(os.Stderr, "FAILED: %s.%s — expected %s, got %s\n", funcName, tcName, expectedISCC, result.Iscc)
			passed = false
		}
	}
	return passed
}

// runAudioTests validates gen_audio_code_v0 conformance vectors.
func runAudioTests(data map[string]map[string]vectorEntry) bool {
	funcName := "gen_audio_code_v0"
	cases, ok := data[funcName]
	if !ok {
		fmt.Fprintf(os.Stderr, "FAILED: %s section missing from conformance data\n", funcName)
		return false
	}

	passed := true
	for tcName, tc := range cases {
		var cvJSON []float64
		if err := json.Unmarshal(tc.Inputs[0], &cvJSON); err != nil {
			fmt.Fprintf(os.Stderr, "FAILED: %s.%s — could not parse test inputs\n", funcName, tcName)
			passed = false
			continue
		}
		cv := make([]int32, len(cvJSON))
		for i, v := range cvJSON {
			cv[i] = int32(v)
		}

		var bits float64
		if err := json.Unmarshal(tc.Inputs[1], &bits); err != nil {
			fmt.Fprintf(os.Stderr, "FAILED: %s.%s — could not parse test inputs\n", funcName, tcName)
			passed = false
			continue
		}

		expectedISCC, _ := tc.Outputs["iscc"].(string)
		result, err := GenAudioCodeV0(cv, uint32(bits))
		if err != nil {
			fmt.Fprintf(os.Stderr, "FAILED: %s.%s — error: %v\n", funcName, tcName, err)
			passed = false
			continue
		}
		if result.Iscc != expectedISCC {
			fmt.Fprintf(os.Stderr, "FAILED: %s.%s — expected %s, got %s\n", funcName, tcName, expectedISCC, result.Iscc)
			passed = false
		}
	}
	return passed
}

// runVideoTests validates gen_video_code_v0 conformance vectors.
func runVideoTests(data map[string]map[string]vectorEntry) bool {
	funcName := "gen_video_code_v0"
	cases, ok := data[funcName]
	if !ok {
		fmt.Fprintf(os.Stderr, "FAILED: %s section missing from conformance data\n", funcName)
		return false
	}

	passed := true
	for tcName, tc := range cases {
		var framesJSON [][]float64
		if err := json.Unmarshal(tc.Inputs[0], &framesJSON); err != nil {
			fmt.Fprintf(os.Stderr, "FAILED: %s.%s — could not parse test inputs\n", funcName, tcName)
			passed = false
			continue
		}
		frameSigs := make([][]int32, len(framesJSON))
		for i, frame := range framesJSON {
			frameSigs[i] = make([]int32, len(frame))
			for j, v := range frame {
				frameSigs[i][j] = int32(v)
			}
		}

		var bits float64
		if err := json.Unmarshal(tc.Inputs[1], &bits); err != nil {
			fmt.Fprintf(os.Stderr, "FAILED: %s.%s — could not parse test inputs\n", funcName, tcName)
			passed = false
			continue
		}

		expectedISCC, _ := tc.Outputs["iscc"].(string)
		result, err := GenVideoCodeV0(frameSigs, uint32(bits))
		if err != nil {
			fmt.Fprintf(os.Stderr, "FAILED: %s.%s — error: %v\n", funcName, tcName, err)
			passed = false
			continue
		}
		if result.Iscc != expectedISCC {
			fmt.Fprintf(os.Stderr, "FAILED: %s.%s — expected %s, got %s\n", funcName, tcName, expectedISCC, result.Iscc)
			passed = false
		}
	}
	return passed
}

// runMixedTests validates gen_mixed_code_v0 conformance vectors.
func runMixedTests(data map[string]map[string]vectorEntry) bool {
	funcName := "gen_mixed_code_v0"
	cases, ok := data[funcName]
	if !ok {
		fmt.Fprintf(os.Stderr, "FAILED: %s section missing from conformance data\n", funcName)
		return false
	}

	passed := true
	for tcName, tc := range cases {
		var codes []string
		if err := json.Unmarshal(tc.Inputs[0], &codes); err != nil {
			fmt.Fprintf(os.Stderr, "FAILED: %s.%s — could not parse test inputs\n", funcName, tcName)
			passed = false
			continue
		}

		var bits float64
		if err := json.Unmarshal(tc.Inputs[1], &bits); err != nil {
			fmt.Fprintf(os.Stderr, "FAILED: %s.%s — could not parse test inputs\n", funcName, tcName)
			passed = false
			continue
		}

		expectedISCC, _ := tc.Outputs["iscc"].(string)
		result, err := GenMixedCodeV0(codes, uint32(bits))
		if err != nil {
			fmt.Fprintf(os.Stderr, "FAILED: %s.%s — error: %v\n", funcName, tcName, err)
			passed = false
			continue
		}
		if result.Iscc != expectedISCC {
			fmt.Fprintf(os.Stderr, "FAILED: %s.%s — expected %s, got %s\n", funcName, tcName, expectedISCC, result.Iscc)
			passed = false
		}
	}
	return passed
}

// decodeStream decodes a "stream:<hex>" value into bytes.
func decodeStream(s string) ([]byte, error) {
	hexData := strings.TrimPrefix(s, "stream:")
	if hexData == "" {
		return []byte{}, nil
	}
	return hex.DecodeString(hexData)
}

// runDataTests validates gen_data_code_v0 conformance vectors.
func runDataTests(data map[string]map[string]vectorEntry) bool {
	funcName := "gen_data_code_v0"
	cases, ok := data[funcName]
	if !ok {
		fmt.Fprintf(os.Stderr, "FAILED: %s section missing from conformance data\n", funcName)
		return false
	}

	passed := true
	for tcName, tc := range cases {
		var streamStr string
		if err := json.Unmarshal(tc.Inputs[0], &streamStr); err != nil {
			fmt.Fprintf(os.Stderr, "FAILED: %s.%s — could not parse test inputs\n", funcName, tcName)
			passed = false
			continue
		}
		inputBytes, err := decodeStream(streamStr)
		if err != nil {
			fmt.Fprintf(os.Stderr, "FAILED: %s.%s — could not decode stream: %v\n", funcName, tcName, err)
			passed = false
			continue
		}

		var bits float64
		if err := json.Unmarshal(tc.Inputs[1], &bits); err != nil {
			fmt.Fprintf(os.Stderr, "FAILED: %s.%s — could not parse test inputs\n", funcName, tcName)
			passed = false
			continue
		}

		expectedISCC, _ := tc.Outputs["iscc"].(string)
		result, err := GenDataCodeV0(inputBytes, uint32(bits))
		if err != nil {
			fmt.Fprintf(os.Stderr, "FAILED: %s.%s — error: %v\n", funcName, tcName, err)
			passed = false
			continue
		}
		if result.Iscc != expectedISCC {
			fmt.Fprintf(os.Stderr, "FAILED: %s.%s — expected %s, got %s\n", funcName, tcName, expectedISCC, result.Iscc)
			passed = false
		}
	}
	return passed
}

// runInstanceTests validates gen_instance_code_v0 conformance vectors.
func runInstanceTests(data map[string]map[string]vectorEntry) bool {
	funcName := "gen_instance_code_v0"
	cases, ok := data[funcName]
	if !ok {
		fmt.Fprintf(os.Stderr, "FAILED: %s section missing from conformance data\n", funcName)
		return false
	}

	passed := true
	for tcName, tc := range cases {
		var streamStr string
		if err := json.Unmarshal(tc.Inputs[0], &streamStr); err != nil {
			fmt.Fprintf(os.Stderr, "FAILED: %s.%s — could not parse test inputs\n", funcName, tcName)
			passed = false
			continue
		}
		inputBytes, err := decodeStream(streamStr)
		if err != nil {
			fmt.Fprintf(os.Stderr, "FAILED: %s.%s — could not decode stream: %v\n", funcName, tcName, err)
			passed = false
			continue
		}

		var bits float64
		if err := json.Unmarshal(tc.Inputs[1], &bits); err != nil {
			fmt.Fprintf(os.Stderr, "FAILED: %s.%s — could not parse test inputs\n", funcName, tcName)
			passed = false
			continue
		}

		expectedISCC, _ := tc.Outputs["iscc"].(string)
		result, err := GenInstanceCodeV0(inputBytes, uint32(bits))
		if err != nil {
			fmt.Fprintf(os.Stderr, "FAILED: %s.%s — error: %v\n", funcName, tcName, err)
			passed = false
			continue
		}
		if result.Iscc != expectedISCC {
			fmt.Fprintf(os.Stderr, "FAILED: %s.%s — expected %s, got %s\n", funcName, tcName, expectedISCC, result.Iscc)
			passed = false
		}
	}
	return passed
}

// runIsccTests validates gen_iscc_code_v0 conformance vectors.
func runIsccTests(data map[string]map[string]vectorEntry) bool {
	funcName := "gen_iscc_code_v0"
	cases, ok := data[funcName]
	if !ok {
		fmt.Fprintf(os.Stderr, "FAILED: %s section missing from conformance data\n", funcName)
		return false
	}

	passed := true
	for tcName, tc := range cases {
		var codes []string
		if err := json.Unmarshal(tc.Inputs[0], &codes); err != nil {
			fmt.Fprintf(os.Stderr, "FAILED: %s.%s — could not parse test inputs\n", funcName, tcName)
			passed = false
			continue
		}

		expectedISCC, _ := tc.Outputs["iscc"].(string)
		// Conformance vectors use default (non-wide) mode
		result, err := GenIsccCodeV0(codes, false)
		if err != nil {
			fmt.Fprintf(os.Stderr, "FAILED: %s.%s — error: %v\n", funcName, tcName, err)
			passed = false
			continue
		}
		if result.Iscc != expectedISCC {
			fmt.Fprintf(os.Stderr, "FAILED: %s.%s — expected %s, got %s\n", funcName, tcName, expectedISCC, result.Iscc)
			passed = false
		}
	}
	return passed
}
